test/versioned/grpc/grpc-server.js

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2022 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+const { ERR_MSG, ERR_CODE } = require('./constants')
+module.exports = function createServerMethods(server) {
+  return {
+    sayHello: function sayHello({ metadata, request: { name } }, cb) {
+      // add the metadata from client that the server receives so we can assert DT functionality
+      server.metadataMap.set(name, metadata.internalRepr)
+      const message = `Hello ${name}`
+      cb(null, { message })
+    },
+    sayHelloClientStream: function sayHelloCStream(call, cb) {
+      const { metadata } = call
+      const names = []
+      call.on('data', function (clientStream) {
+        const { name } = clientStream
+        server.metadataMap.set(name, metadata.internalRepr)
+        names.push(name)
+      })
+      call.on('end', function () {
+        cb(null, {
+          message: `Hello ${names.join(', ')}`
+        })
+      })
+    },
+    sayHelloServerStream: function sayHelloCStream(call) {
+      const {
+        metadata,
+        request: { name }
+      } = call
+      name.forEach((n) => {
+        // add the metadata from client that the server receives so we can assert DT functionality
+        server.metadataMap.set(n, metadata.internalRepr)
+        call.write({ message: `Hello ${n}` })
+      })
+      call.end()
+    },
+    sayHelloBidiStream: function sayHelloCStream(call) {
+      const { metadata } = call
+      call.on('data', (clientStream) => {
+        const { name } = clientStream
+        // add the metadata from client that the server receives so we can assert DT functionality
+        server.metadataMap.set(name, metadata.internalRepr)
+        call.write({ message: `Hello ${name}` })
+      })
+      call.on('end', () => {
+        call.end()
+      })
+    },
+    sayError: function sayError(whatever, cb) {
+      return cb({
+        code: ERR_CODE,
+        message: ERR_MSG
+      })
+    },
+    sayErrorClientStream: function sayErrorClientStream(call, cb) {
+      call.on('data', function () {
+        // no-op as we do not care about the data
+      })
```

Comment:
I'm just wondering if our instrumentation could somehow break if we encounter errors in the middle of streaming.

Reply:
Ah. I'm not sure as we're just implementing the onReceiveStatus hook.  I could try testing but it may be more relevant to the server instrumentation which we will build later?

Reply:
so I think what you would do is add an error handler.

```call.on('error'), (err) => {
 // handle server error
})
```

Reply:
actually that'd be an error from the client stream, 🤔

Reply:
Ok so we would handle it.  So if i changed the code to do something like this

```
    sayErrorClientStream: function sayErrorClientStream(call, cb) {
      call.on('data', function (stream) {
        if (stream.name === 'error') {
          cb({
            code: ERR_CODE,
            message: 'bob test'
          })
        }

        // no-op as we do not care about the data
      })

      call.on('end', function () {
        cb({
          code: ERR_CODE,
          message: ERR_MSG
        })
      })
    },
```

It would capture the error accordingly so we would handle this.

Reply:
let me add another test to handle this

Reply:
k, done!